File: verilog/dv/cocotb/tests/user_project/user_address_space.py
import random
import cocotb
from cocotb.triggers import FallingEdge,RisingEdge,ClockCycles,Timer,Edge
import cocotb.log
from interfaces.cpu import RiskV
from interfaces.defsParser import Regs
from cocotb.result import TestSuccess
from tests.common_functions.test_functions import *
from tests.bitbang.bitbang_functions import *
from interfaces.caravel import GPIO_MODE
from interfaces.common import Macros


bit_time_ns = 0
reg = Regs()


@cocotb.test()
@repot_test
async def user_address_space(dut):
    caravelEnv,clock = await test_configure(dut,timeout_cycles=17936)
    cocotb.log.info(f"[TEST] Start user_address_space test")
    ack_hdl  = caravelEnv.caravel_hdl.mprj.addr_space_testing.wbs_ack_o 
    addr_hdl  = caravelEnv.caravel_hdl.mprj.addr_space_testing.addr 
    data_hdl  = caravelEnv.caravel_hdl.mprj.addr_space_testing.data 
    addr_arr = [0x30000000,0x30000004,0x30000008,0x300FFFF8,0x300FFFFC,0x30100000,0x30044078,0x30090A94,0x3005FE2C,0x300F9F44,0x30032E58,0x300602EC,0x30100000]
    data_arr = [0x97cf0d2d,0xbc748313,0xbfda8146,0x5f5e36b1,0x0c1fe9d9,0x6d12d2b8,0xdcd244d1,0x0da37088,0x7b8e4345,0x00000777,0x00000777,0x00000777,0xFFFFFFFF]
    for addr, data in zip(addr_arr, data_arr):
        await RisingEdge(ack_hdl)
        if addr_hdl.value.integer != addr:
            cocotb.log.error(f"[TEST] seeing unexpected address {hex(addr_hdl.value.integer)} expected {hex(addr)}")
        elif data_hdl.value.integer !=  data:
            cocotb.log.error(f"[TEST] seeing unexpected data {hex(data_hdl.value.integer)} expected {hex(data)} address {hex(addr)}")
        else: 
            cocotb.log.info(f"[TEST] seeing the correct data {hex(data)} from address {hex(addr)}")

File: verilog/dv/cocotb/tests/user_project/user_address_space.c

#include <defs.h>
#include <stub.c>

// --------------------------------------------------------

void main()
{
    int j;
    reg_wb_enable =1; // for enable writing to user space area
    // space from 0x30000000 to 0x30100000 according to current linker script 
    // TODO: fix when issue #107 https://github.com/efabless/caravel_mgmt_soc_litex/issues/107 is resolved

    // first 3 addresses 
    (*(volatile uint32_t*) 0x30000000 ) = 0x97cf0d2d; 
    (*(volatile uint32_t*) 0x30000004 ) = 0xbc748313; 
    (*(volatile uint32_t*) 0x30000008 ) = 0xbfda8146; 

    // last 3 addresses 
    (*(volatile uint32_t*) 0x300FFFF8 ) = 0x5f5e36b1; 
    (*(volatile uint32_t*) 0x300FFFFC ) = 0x0c1fe9d9; 
    (*(volatile uint32_t*) 0x30100000 ) = 0x6d12d2b8; 


    // random addresses inside the user space 
    (*(volatile uint32_t*) 0x30044078 ) = 0xdcd244d1; 
    (*(volatile uint32_t*) 0x30090A94 ) = 0x0da37088; 
    (*(volatile uint32_t*) 0x3005FE2C ) = 0x7b8e4345; 
    // random read 
    int temp;
    temp = (*(volatile uint32_t*) 0x300F9F44 ); 
    temp = (*(volatile uint32_t*) 0x30032E58 ); 
    temp = (*(volatile uint32_t*) 0x300602EC ); 

    // addresses outside user space - injecting error if user project ack is affected
    reg_mprj_io_15 = GPIO_MODE_MGMT_STD_OUTPUT;
    reg_mprj_io_14 = GPIO_MODE_MGMT_STD_OUTPUT;
    
    // finish with writing last address with Fs
    (*(volatile uint32_t*) 0x30100000 ) = 0xFFFFFFFF; 


}
